arangod/Pregel/Actor/include/Actor/Actor.h

```diff
@@ -0,0 +1,248 @@
+      if (payload.ok()) {
+        runtime->dispatch(pid, sender, payload.get());
+      } else {
+        fmt::print("Error serializing UnknownMessage");
+        std::abort();
+      }
+    }
+  }
+
+  auto finish() -> void override { finished.store(true); }
+  auto finishedAndIdle() -> bool override {
+    return finished.load() and idle.load();
+  }
+
+  auto serialize() -> velocypack::SharedSlice override {
+    auto res = inspection::serializeWithErrorT(*this);
+    if (not res.ok()) {
+      std::abort();
+    }
+    return res.get();
+  }
+
+ private:
+  void push(ActorPID sender, typename Config::Message&& msg) {
+    pushToQueueAndKick(std::make_unique<InternalMessage>(
+        sender,
+        std::make_unique<MessageOrError<typename Config::Message>>(msg)));
+  }
+  void push(ActorPID sender, ActorError&& msg) {
+    pushToQueueAndKick(std::make_unique<InternalMessage>(
+        sender,
+        std::make_unique<MessageOrError<typename Config::Message>>(msg)));
+  }
+
+  void kick() {
+    // Make sure that *someone* works here
+    (*runtime->scheduler)([self = this->weak_from_this()]() {
+      auto me = self.lock();
+      if (me != nullptr) {
+        me->work();
+      }
+    });
+  }
+
+  void work() {
+    if (finished.load() == true) {
+      return;
+    }
+    auto i = batchSize;
+
+    while (auto msg = inbox.pop()) {
+      state = std::visit(
+          typename Config::template Handler<Runtime>{
+              {pid, msg->sender, std::move(state), runtime}},
+          msg->payload->item);
+      i--;
+      if (i == 0) {
+        // push more work to scheduler if queue is still not empty
+        if (not inbox.empty()) {
+          kick();
+          return;
+        }
+        break;
+      }
+    }
+
+    idle.store(true);
+
+    // push more work to scheduler if a message was added to queue after
+    // previous inbox.empty check
+    auto isIdle = true;
+    if (not inbox.empty() and idle.compare_exchange_strong(isIdle, false)) {
+      kick();
+    }
+  }
+
+  struct InternalMessage
+      : arangodb::pregel::mpscqueue::MPSCQueue<InternalMessage>::Node {
+    InternalMessage(
+        ActorPID sender,
+        std::unique_ptr<MessageOrError<typename Config::Message>>&& payload)
+        : sender(sender), payload(std::move(payload)) {}
+    ActorPID sender;
+    std::unique_ptr<MessageOrError<typename Config::Message>> payload;
+  };
+  template<typename Inspector>
+  auto inspect(Inspector& f, InternalMessage& x) {
+    return f.object(x).fiels(f.field("sender", x.sender),
+                             f.field("payload", x.payload));
+  }
+
+  auto pushToQueueAndKick(std::unique_ptr<InternalMessage> msg) -> void {
+    inbox.push(std::move(msg));
+
+    // only push work to scheduler if actor is idle (meaning no work is waiting
+    // on the scheduler and no work is currently processed in work())
+    auto isIdle = idle.load();
+    if (isIdle and idle.compare_exchange_strong(isIdle, false)) {
```

Comment:
Now you kick almost always, even in case when idle was false.

Btw do you understand what with idle you just write spinlock with try_lock and unlock?
Please specify memory_orders

```suggestion
    auto isIdle = true
    if (idle.load() == isIdle and idle.compare_exchange_strong(isIdle, false)) {
```

Reply:
Your suggestion is semantically equivalent.
```
    auto isIdle = idle.load();
    if (isIdle and idle.compare_exchange_strong(isIdle, false)) {
```
We first load the value of `idle` and store it in `isIdle`, then we check that `isIdle` (the local variable) is true, and only if that is the case we try the compare_exchange.

Where do you see a spinlock? With the idle flag just ensure that only a single worker can be enqueued at any point in time.

We deliberately did not specify any memory orders yet since this requires careful analysis.

Reply:
> Your suggestion is semantically equivalent.

You're right I misread.

> Where do you see a spinlock? With the idle flag just ensure that only a single worker can be enqueued at any point in time.

Here is try_lock, when we store false it's unlock.

> We deliberately did not specify any memory orders yet since this requires careful analysis.

It's devel, we can use simpler solution, because here we have really tricky memory orders (correct for queue and correct for spinlock, don't work together)

Reply:
> Here is try_lock, when we store false it's unlock.
I don't get it... there is not loop involved and we are not spinning on anything. This compare_exchange is simply used to ensure that at most one thread is scheduling a worker, nothing more.

Reply:
I didn't write somewhere here is loop.

I wrote here is try_lock (just cas), and in the another place is unlock (store false)

In general nevermind it's not the issue but you can separate this part and make it more clear

Reply:
I don't understand why you interpret this as lock/unlock, but I wanted to add some comments anyway.

Reply:
I don't wrote "lock"

```
class SpinLock {
  public:
    bool try_lock() {
       auto locked = _locked.load(relaxed)
       return !locked && _locked.cas(locked, false, acquire, relaxed);
    }
    void unlock() noexcept {
       locked.store(release);
    }
  private:
    std::atomic_bool _locked{false};
};
```

It's spinlock, in this code it's same.
As I already wrote it's ok, in general we are protecting from multiple consumers of queue